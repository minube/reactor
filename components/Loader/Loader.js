import { bool, string } from 'prop-types';
import { Animated, View } from 'react-native';
import React, { PureComponent } from 'react';
import Text from '../Text';
import styles from './Loader.style';


export default class Loader extends PureComponent {
  static propTypes = {
    color: string,
    large: bool,
    text: string,

  };

  static defaultProps = {
    color: undefined,
    large: false,
    text: undefined,
  };

  state = {
    dotsOpacities: [new Animated.Value(0), new Animated.Value(0), new Animated.Value(0)],
    duration: 300,
  };

  constructor(props) {
    super(props);

    this.animationState = {
      targetOpacity: 1,
      animated: true,
    };
  }

  componentDidMount() {
    const { _animation } = this;
    _animation(0);
  }

  componentWillUnmount() {
    this.animationState.animated = false;
  }

  _animation = (initialDot) => {
    const { _animation, animationState, state: { dotsOpacities, duration } } = this;

    if (!animationState.animated) return;
    const min = 0;
    let dot = initialDot;

    if (dot >= dotsOpacities.length) {
      dot = 0;
      animationState.targetOpacity = animationState.targetOpacity === min ? 1 : min;
    }

    const nextDot = dot + 1;

    Animated.timing(dotsOpacities[dot], {
      toValue: animationState.targetOpacity,
      duration,
    }).start(() => _animation(nextDot));
  }

  render() {
    const {
      props: {
        text, color, large,
      },
      state: { dotsOpacities },
    } = this;

    return (
      <View style={styles.container}>
        { text && (
          <Text
            color={color}
            style={[
              styles.regular,
              large && styles.large,
            ]}
          >
            {text}
          </Text>
        )}
        { dotsOpacities.map(dotOpacity => (
          <Animated.Text style={[{ opacity: dotOpacity }]}>
            <Text>{' '}</Text>
            <Text
              color={color}
              style={[
                styles.regular,
                large && styles.large,
              ]}
            >
            .
            </Text>
          </Animated.Text>
        ))}
      </View>
    );
  }
}
