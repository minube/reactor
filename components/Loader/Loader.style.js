import { StyleSheet } from 'react-native';

import { THEME } from '../../common';

const { UNIT, FONT } = THEME;
const DOT_SIZE = UNIT * 3.2;

export default StyleSheet.create({
  container: {
    flexDirection: 'row',
  },

  dot: {
    transform: [
      { translateY: -DOT_SIZE * 0.2 },
    ],
  },

  dotRegular: {
    ...FONT.BUTTON,
    fontSize: DOT_SIZE,
  },

  textRegular: {
    ...FONT.BUTTON,
    fontSize: UNIT * 1.6,
  },

  dotLarge: {
    fontSize: UNIT * 5.4,
    letterSpacing: UNIT * 0.12,
  },
});
