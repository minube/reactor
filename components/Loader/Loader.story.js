import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, color, select } from '@storybook/addon-knobs/react';

import Loader from './Loader';

storiesOf('✅ Loader', module)
  .add('default', () => (
    <Loader />
  ))
  .add('text', () => (
    <Loader text="Loading" />
  ))
  .add('🏀 Playground', () => (
    <Loader
      animating={boolean('animating', true)}
      color={color('color', 'green')}
      hidesWhenStopped={boolean('hidesWhenStopped', false)}
      size={select('size', ['small', 'large'], 'small')}
    />
  ));
