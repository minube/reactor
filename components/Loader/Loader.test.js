import React from 'react';
import renderer from 'react-test-renderer';

import Loader from './Loader';

const COLOR = 'red';
const LOADING = 'Loading';

jest.useFakeTimers()

describe('<Loader>', () => {
  it('renders', () => {
    const tree = renderer.create(<Loader />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {color}', () => {
    const tree = renderer.create(<Loader color={COLOR} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {large}', () => {
    const tree = renderer.create(<Loader large />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {text}', () => {
    const tree = renderer.create(<Loader text={LOADING} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
