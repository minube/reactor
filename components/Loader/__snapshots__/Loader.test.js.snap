// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader> renders 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
</View>
`;

exports[`<Loader> when {color} 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "color": "red",
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "color": "red",
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "color": "red",
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
</View>
`;

exports[`<Loader> when {large} 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 54,
            "fontWeight": "600",
            "letterSpacing": 1.2,
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 54,
            "fontWeight": "600",
            "letterSpacing": 1.2,
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 54,
            "fontWeight": "600",
            "letterSpacing": 1.2,
          },
        ]
      }
    >
      .
    </Text>
  </Text>
</View>
`;

exports[`<Loader> when {text} 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#4a4a4a",
          "fontFamily": undefined,
          "fontWeight": "400",
          "textAlignVertical": "center",
        },
        Object {
          "fontSize": 16,
          "lineHeight": 30,
        },
        false,
        Object {
          "fontFamily": undefined,
          "fontSize": 16,
          "fontWeight": "600",
        },
      ]
    }
  >
    Loading
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "translateY": -6.4,
          },
        ],
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {},
        ]
      }
    >
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#4a4a4a",
            "fontFamily": undefined,
            "fontWeight": "400",
            "textAlignVertical": "center",
          },
          Object {
            "fontSize": 16,
            "lineHeight": 30,
          },
          false,
          Object {
            "fontFamily": undefined,
            "fontSize": 32,
            "fontWeight": "600",
          },
        ]
      }
    >
      .
    </Text>
  </Text>
</View>
`;
